### PR TITLE
Upgrade to Lucene 4.10.2

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @ymatsuda @eishay

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/engine/query/KBoostQuery.scala [hunk 1]
```diff
@@ -77,7 +77,7 @@ class KBoostWeight(override val query: KBoostQuery, override val searcher: Index
     result
   }
 
-  override def scorer(context: AtomicReaderContext, scoreDocsInOrder: Boolean, topScorer: Boolean, acceptDocs: Bits): Scorer = {
+  override def scorer(context: AtomicReaderContext, acceptDocs: Bits): Scorer = {
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/engine/query/KBoostQuery.scala:80]: @ymatsuda the new `Weigth.scorer` API gets rid of `scoreDocsInOrder` and `topScorer`, my understanding is that this is fine, even for queries that override `scoresDocsOutOfOrder` to `false`, could you confirm?

File: kifi-backend/modules/search/app/com/keepit/search/index/DefaultAnalyzer.scala [hunk 2]
```diff
@@ -98,7 +97,7 @@ object DefaultAnalyzer {
     "ro" -> stdAnalyzer.withFilter[LowerCaseFilter].withStopFilter(_.Romanian),
     "ru" -> stdAnalyzer.withFilter[LowerCaseFilter].withStopFilter(_.Russian),
     "sv" -> stdAnalyzer.withFilter[LowerCaseFilter].withStopFilter(_.Swedish),
-    "th" -> stdAnalyzer.withFilter[LowerCaseFilter].withFilter[ThaiWordFilter].withStopFilter(_.Thai),
+    "th" -> thAnalyzer.withFilter[LowerCaseFilter].withStopFilter(_.Thai),
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/DefaultAnalyzer.scala:100]: @ymatsuda  `ThaiWordFilter` has been deprecated.

File: kifi-backend/modules/search/app/com/keepit/search/index/DefaultAnalyzer.scala [hunk 3]
```diff
@@ -150,18 +149,12 @@ class Analyzer(tokenizerFactory: TokenizerFactory,
...
-          val constructor = m.runtimeClass.getConstructor(classOf[TokenStream]).asInstanceOf[Constructor[TokenStream]]
-          withFilter(WrapperTokenFilterFactory(constructor))
-        } catch {
-          case ex: NoSuchMethodException =>
-            log.error("failed to find a filter constructor: %s".format(m.runtimeClass.toString))
-            this
-        }
+        log.error("failed to find a filter constructor: %s".format(m.runtimeClass.toString))
+        this
     }
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/DefaultAnalyzer.scala:158]: @ymatsuda also see `WrapperTokenFilterFactory`

File: kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala [hunk 4]
```diff
@@ -190,10 +190,11 @@ abstract class Indexer[T, S, I <: Indexer[T, S, I]](
...
-  def indexSize = {
-    Some(searcher.indexReader.inner.getIndexCommit().getFileNames().map { filename =>
-      if (indexDirectory.fileExists(filename)) indexDirectory.fileLength(filename) else 0L
-    }.sum)
+  def indexSize = indexDirectory.asFile().map(_.listFiles().toSet).map { directoryFiles =>
+    searcher.indexReader.inner.getIndexCommit().getFileNames().collect {
+      case filename if directoryFiles.contains(filename) =>
+        indexDirectory.fileLength(filename)
+    }.sum
   }
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala:198]: @ymatsuda `Directory.fileExists` has been deprecated so I did this in a couple of places.

**Comment by eishay** [comment]
wohooo

File: kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala [hunk 5]
```diff
@@ -135,9 +132,6 @@ object WrapperTokenFilterFactory {
   def apply(constructor: Constructor[TokenStream]) = new TokenFilterFactory {
     def apply(tokenStream: TokenStream) = constructor.newInstance(tokenStream)
   }
-  def apply(constructor: Constructor[TokenStream], version: Version) = new TokenFilterFactory {
-    def apply(tokenStream: TokenStream) = constructor.newInstance(version, tokenStream)
-  }
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala:140]: @ymatsuda I could get rid of this

File: kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala [hunk 6]
```diff
@@ -148,8 +142,7 @@ class FrenchElisionFilter(tokenStream: TokenStream)
 class SymbolDecompounder(tokenStream: TokenStream) extends TokenFilter(tokenStream) {
   val termAttr = addAttribute(classOf[CharTermAttribute])
   val posIncrAttr = addAttribute(classOf[PositionIncrementAttribute])
-  val typeAttr = tokenStream.getAttribute(classOf[TypeAttribute]).asInstanceOf[TypeAttributeImpl]
-  val tokenType = new TypeAttributeAccessor
+  val tokenAttr = tokenStream.getAttribute(classOf[TypeAttribute]).asInstanceOf[PackedTokenAttributeImpl]
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala:145]: @ymatsuda I switched `TypeAttributeImpl` with `PackedTokenAttributeImpl`, that seems to have done the trick.

File: kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala [hunk 7]
```diff
@@ -239,14 +232,3 @@ class EmptyTokenFilter(tokenStream: TokenStream) extends TokenFilter(tokenStream
...
-class TypeAttributeAccessor extends TypeAttributeImpl {
-  var tokenType: String = TypeAttribute.DEFAULT_TYPE
-  override def setType(tt: String) { tokenType = tt }
-
-  def apply(ta: TypeAttributeImpl) = {
-    ta.copyTo(this)
-    tokenType
-  }
-}
-
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/TokenFilterFactory.scala:252]: @ymatsuda and also got rid of this using the `type` method on `PackedTokenAttributeImpl`

File: kifi-backend/modules/search/test/com/keepit/search/query/ProximityQueryTest.scala [hunk 8]
```diff
@@ -21,7 +21,7 @@ import com.keepit.search.SearchConfig
 
 class ProximityQueryTest extends Specification {
 
-  val config = new IndexWriterConfig(Version.LUCENE_47, DefaultAnalyzer.defaultAnalyzer)
+  val config = new IndexWriterConfig(Version.LATEST, DefaultAnalyzer.defaultAnalyzer)
```
Comment (leogrim) [inline kifi-backend/modules/search/test/com/keepit/search/query/ProximityQueryTest.scala:24]: @ymatsuda

**Comment by ymatsuda** [comment]
Thanks! :+1:  :+1:

**Comment by stkem** [comment]
:-)

**Pull request merged**